docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,224 @@
+
+```sql
+SET GLOBAL tidb_enable_legacy_instance_scope = 1;
+```
+
+We can default to `TRUE` for now, but should revisit this in the future.
+
+### Stage 2: Mapping All Configuration File Settings to System Variables
+
+The second stage is to map all configuration file settings to system variable names. In MySQL any configuration file setting is also available as a system variable (even if it is read only), which makes it possible to get the configuration of the cluster with `SHOW GLOBAL VARIABLES`. Once this step is complete, it should also be possible to offer an `[instance]` section of the configuration file which accepts the system variable names (and finally consistent naming between the two systems). We can then modify the example configuration files to be based on the flat `[instance]` configuration:
+
+```
+# tidb.toml
+[instance]
+max_connections = 1234
+tidb_general_log = "/path/to/file"
+```
+
+The default will be that we map variables as read-only, so we don't need to think about the specific cases (i.e. you can't easily change the socket or port). This helps us achieve completeness first, and then we can then evaluate which ones can be made dynamic.
+
+### Stage 3: Refactoring
+
+The use of a `GetGlobal()` and `SetGlobal()` func for each instance scoped system variable is not ideal. It is possible to refactor the system variable framework so that instance scope is stored in a map, and the values are updated automatically by `SET GLOBAL` on an instance scoped variable. On startup, as the configuration file is parsed it will update the values in the map. This seems like a better approach than the current use of Setters/Getters, and because there is a prescribed way of doing it we can correctly handle the data races that are common with our current incorrect usage of calling `config.GetGlobalConfig()`.
+
+Thus, the source of truth for instance scoped variables moves from the `config` package to another part of the server (likely `domain`). See [issue #30366](https://github.com/pingcap/tidb/issues/30366).
+
+## Impacts & Risks
+
+The biggest risk is that we can not agree on a reduced scope of implementation and the project extends to cover increased scope. Configuration management is a classic example of a [bikeshed problem](https://en.wikipedia.org/wiki/Law_of_triviality), and we will likely need to make some tradeoffs to get anywhere.
+
+Many of the alternatives are difficult to implement because they break compatibility (we need to live with `GLOBAL` means cluster `GLOBAL`, and there is no `SET CLUSTER`) or they break rolling upgrade scenarios because we currently have no specific rules of the minimum version of TiDB which can be upgraded from.
+
+## Investigation & Alternatives
+
+- There is no equivalent functionality to reference in MySQL since it does not have the native concept of a cluster.
```

Comment:
MySQL does have the concept of a cluster: InnoDB Cluster / Group Replication. And NDB is also a cluster, but with that the cluster bits are mostly outside of `mysqld`.

https://dev.mysql.com/doc/mysql-shell/8.0/en/configuring-innodb-cluster.html shows `Cluster.setInstanceOption(instance, option, value)` which could be related to this PR, but this is done via MySQL Shell.

Reply:
I could improve the wording, but InnoDB Cluster and NDB do not provide a native concept *in configuration* context.